src/system/SystemObject.h

```diff
@@ -228,10 +304,32 @@ template <class T, unsigned int N>
 inline T * ObjectPool<T, N>::TryCreate(Layer & aLayer)
 {
     T * lReturn = nullptr;
-    unsigned int lIndex;
-#if CHIP_SYSTEM_CONFIG_PROVIDE_STATISTICS
-    unsigned int lNumInUse = 0;
-#endif
+
+#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+    T * newNode = new T();
+
+    if (newNode->TryCreate(aLayer, sizeof(T)))
+    {
+        std::lock_guard<std::mutex> lock(mMutex);
+        Object * p = &mDummyHead;
+
+        // Traverse down to the end of the list
+        while (p->next)
+        {
+            p = p->next;
+        }
+
+        p->next            = newNode;
+        newNode->prev      = p;
+        newNode->mMutexRef = &mMutex;
+        lReturn            = newNode;
```

Comment:
You can O(1) insert at the head of the list:

```suggestion
        Object * p = &mDummyHead;
        if (p->next)
        {
            p->next->prev =  newNode;
        }
        newNode->next = p->next;
        p->next            = newNode;
        newNode->prev      = p;
        newNode->mMutexRef = &mMutex;
        lReturn            = newNode;
```